### PR TITLE
Merge 2.9 develop

### PR DESCRIPTION
**Pull request by SimonRichardson**
This is a merge from 2.9 into develop.

This wasn't a clean merge:
 - version numbers conflict: moved to 3.0-beta1
 - conflict around where caas/metadata.go file was located, this wasn't picked up in the merge but a build after-wards.

c1cb4aab4d (upstream/2.9, origin/2.9, 2.9) Merge pull request #12471 from wallyworld/merge-2.8-20201217a
551e1be182 Merge pull request #12469 from jujubot/increment-to-2.9-rc4
c5305d8fbe Merge pull request #12468 from wallyworld/merge-2.8-20201217
0f9a02959a Merge pull request #12465 from SimonRichardson/charmhub-hide-local-schema-in-errors-messages
4b2d0f8bf8 Merge pull request #12271 from SimonRichardson/wait-for-summary
2002602a86 Merge pull request #12461 from SimonRichardson/charmhub-bundle-descriptions
b4037cb9ec Merge pull request #12460 from hmlanigan/bundle-output
97933e4090 Merge pull request #12455 from SimonRichardson/charmhub-ensure-charmorigin-output
9f56bc9f7f Merge pull request #12457 from hmlanigan/unit-resource-doc-not-local
9c119e3c7e (manadart/2.9) Merge pull request #12448 from SimonRichardson/acceptance-tests-architecture
5b260dbbb2 Merge pull request #12449 from hmlanigan/charmhub-no-all
236efd7279 Merge pull request #12458 from hmlanigan/identify-charms-deployed
6bd766cf37 Merge pull request #12456 from jameinel/2.8-into-2.9

### CODE REVIEW

File: apiserver/facades/client/charms/client.go [hunk 1]
```diff
@@ -276,7 +276,7 @@ func normalizeCharmOrigin(origin params.CharmOrigin) (params.CharmOrigin, error)
 	var os string
 	var oSeries string
 	if origin.Series == "all" {
-		logger.Tracef("Series all detected, removing all from the origin. %s", origin.ID)
+		logger.Warningf("Series all detected, removing all from the origin. %s", origin.ID)
```
Comment (achilleasa) [inline apiserver/facades/client/charms/client.go:279]: Should all be quoted to avoid confusion?

File: apiserver/facades/client/charms/client.go [hunk 2]
```diff
@@ -287,12 +287,15 @@ func normalizeCharmOrigin(origin params.CharmOrigin) (params.CharmOrigin, error)
...
 		os = strings.ToLower(sys.String())
+
 		oSeries = origin.Series
 	}
 
 	var arc string
 	if origin.Architecture != "all" {
 		arc = origin.Architecture
+	} else {
+		logger.Warningf("Architecture all detected, removing all from the origin. %s", origin.ID)
```
Comment (achilleasa) [inline apiserver/facades/client/charms/client.go:298]: Likewise here

File: caas/kubernetes/provider/pod.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+	"time"
+
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/status"
+
+	core "k8s.io/api/core/v1"
+)
+
+type EventGetter func() ([]core.Event, error)
```
Comment (achilleasa) [inline caas/kubernetes/provider/pod.go:17]: Exported types should have comments.

File: caas/kubernetes/provider/pod.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+const (
```
Comment (achilleasa) [inline caas/kubernetes/provider/pod.go:19]: Exported types should have comments or a single comment above `const`

File: cmd/plugins/juju-wait-for/strategy.go [hunk 5]
```diff
@@ -195,3 +196,40 @@ func (m *GenericScope) Clone() query.Scope {
...
 	}
 }
+
+// ScopeContext defines a context for a given scope.
+type ScopeContext struct {
+	idents   set.Strings
+	children map[string]map[string]ScopeContext
+}
+
+func MakeScopeContext() ScopeContext {
```
Comment (achilleasa) [inline cmd/plugins/juju-wait-for/strategy.go:206]: Exported types should have a comment

File: core/charm/strategies.go [hunk 6]
```diff
@@ -279,7 +280,8 @@ func (p *Strategy) normalizePlatform(platform Platform) (Platform, error) {
 		os = strings.ToLower(sys.String())
 	}
 	arc := platform.Architecture
-	if platform.Architecture == "" {
+	if platform.Architecture == "" || platform.Architecture == "all" {
+		p.logger.Warningf("Received charm Architecture: %q, changing to %q, for charm %q", platform.Architecture, arch.DefaultArchitecture, p.charmURL)
```
Comment (achilleasa) [inline core/charm/strategies.go:284]: Shouldn't the warning include the reason for the change?

File: resource/repositories/repository.go [hunk 7]
```diff
@@ -52,18 +54,24 @@ func (cfo operationsRepository) get(name string) (resource.Resource, io.ReadClos
...
 	if cfo.repo == nil {
 		res := resource.Resource{
 			Resource: chRes,
 		}
 		return res, reader, nil // a no-op
 	}
-	defer func() { _ = reader.Close() }()
+	defer func() {
+		if err != nil {
+			// With no err, the reader was closed down in unitSetter Read().
```
Comment (achilleasa) [inline resource/repositories/repository.go:69]: This is an ugly hack. We should change `Read` to accept an `io.Reader` and have the caller deal with closing it (assuming it is a reader that can be closed and not a StringReader or something like that)

File: state/cleanup.go [hunk 8]
```diff
@@ -313,7 +315,17 @@ func (st *State) cleanupForceDestroyedRelation(prefix string) (err error) {
 			return errors.Annotatef(err, "leaving scope for unit %q in relation %q", unitName, relation)
 		}
 	}
-
+	if !haveRelationUnits {
+		// We got here because a relation claimed to have units but
+		// there weren't any corresponding relation unit records.
+		// We know this should be forced, and we've already waited the
+		// required time.
+		errs, err := relation.DestroyWithForce(true, 0)
```
Comment (achilleasa) [inline state/cleanup.go:323]: This is an interesting return tuple...

**Review by achilleasa — APPROVED** [review]
Even though this is a merge, I left a few comments that we should probably address in the future.

**Comment by SimonRichardson** [comment]
$$merge$$

File: resource/repositories/repository.go (hunk 7 above)
Comment (SimonRichardson) [reply resource/repositories/repository.go:69]: The io.Reader needs to be re-worked here, it leaks out everywhere!

**Comment by wallyworld** [comment]
!!build!!

**Pull request merged**